Sources/LeafKit/LeafRenderer.swift

```diff
@@ -176,65 +274,87 @@ public final class LeafRenderer {
         )
         return try serializer.serialize()
     }
-    
-    private func expand(path: String) -> String {
-        var path = path
+
+    private func expand(template: String) -> String {
+        var path = template
         // ignore files that already have a type
-        if path.split(separator: ".").count < 2, !path.hasSuffix(".leaf") {
+        if path.split(separator: "/").last?.split(separator: ".").count ?? 1 < 2  , !path.hasSuffix(".leaf") {
```

Comment:
I am 100% not tied to this, just bashed my head against the wall for a minute when my test templates were being named templateX.Y.Z.leaf to indicate layer and couldn't figure out whey it was throwing string errors nonstop. Just consider it a minor improvement.

And I'm not sure I'd WANT files starting with a period to be readable, personally. Don't want a user-generated template to be able to import hidden control files if they're in the directory.

Reply:
Well, dot-starting filenames are an open question, sure - but that should be checked at some layer above this, not blocked here as a side-effect of the particular "does it already have a `.leaf` extension" detection.

Actually, come to think of it, the whole thing is questionable behavior to put down at this layer either, for that matter; it should arguably be up to the particular `LeafFiles` implementation to decide whether to magically look for extensions.

Reply:
No disagreements - after all, I did not write this, just wanted to make it slightly less aggressive! Definitely could use improvement.

Though I'd also find it annoying to have to refer to templates as `render("template.leaf")` nonstop and I'd consider that breaking. Possibly it should just be the other way around though? If RENDER is given a template name that has an extension `render("template.html")` it skips adding `.leaf`?

Though it also currently allows accessing files outside the configured directory when fully pathed, and while removing that would be breaking, maybe not great security wise in the first place?

Reply:
I think the extension-handling behavior should live in the `NIOLeafFiles` implementation that most people end up with by default anyway (and that that implementation should also be responsible for ensuring the realpath of the template after all manipulations have been applied has a common prefix with the configured root directory, something vaguely like:
```swift
// URL canonicalizes relative paths, quick and dirty way to realpath() a path
let rootDirectoryCanonical = URL(fileURLWithPath: rootDirectory, isDirectory: true).path
var templateURL = URL(fileURLWithPath: path, isDirectory: false) 
if templateURL.pathExtension.isEmpty {
    templateURL.appendPathExtension("leaf")
}
// Check the canonical root directory is a prefix of the final template URL
// I mean, this is basically only as good as PHP's crummy "open_basedir" nonsense,
// it won't even come close to stopping every possible hack for getting at paths outside
// the defined root. It does stop all the easy ways though, and hopefully people aren't
// passing user input directly in as a Leaf render path anyway...
guard templateURL.path.hasPrefix(rootDirectoryCanonical) else {
    throw Abort(.forbidden) // or something
}
// go on to read from templateURL.path
```

The overall point being this logic belongs on implementations of `LeafFiles`, and most definitely _not_ on `LeafRenderer`.

Reply:
(Also, putting it up there will make importing Foundation to use for path handling less painful. It can be safely done without using `URL`, but then you have to import the platform-specific Libc, cope with the funny semantics of `realpath()`, and basically you still (probably) won't come up with something that covers as many cases as letting URL clean up the path in one go, not to mention being as bug-free.)

Reply:
Yeah, I'm just not trying to focus too much energy on this for *this* PR since I think more discussion is needed on what expectations are, and contrary to the size of this PR, I was trying to *avoid* reinventing the wheel if I could - definitely think it bears further discussion and addressing, though.

Reply:
Another one for the todo list! 🙃